user-guide/modules/ROOT/pages/advanced-builds.adoc

```diff
@@ -1,3 +1,247 @@
+
+== Custom Build Steps
+
+Your application might require pre-processing steps before compiling. For instance, you might need to generate some pass:[C++] code based on a configuration file. Custom build steps in Boost.Build can be achieved using _rule definitions_ and _actions_. Rules in Boost.Build are a little like functions in a programming language. You can define your own rule to execute custom commands, and then invoke the rule at the appropriate point in your Jamfile.
+
+Let's consider a scenario where you need to generate a pass:[C++] header file from an XML file before compiling the rest of your source code.
+
+. First, define a rule, in your Jamfile, to describe the transformation. The rule takes an input XML file and produces an output header file:
++
+[source,txt]
+----
+rule GenerateHeader ( xmlFile : hppFile )
+{
+    IMPORT $(xmlFile) : generateHeader : : generate ;
+    generate $(xmlFile) : $(hppFile) ;
+}
+----
++
+This rule imports a Python script (generateHeader.py) that transforms an XML file into a pass:[C++] header file. The `IMPORT` command makes a rule from the Python script available in the Jamfile, and generate is the Python rule responsible for transforming the XML file into the header file.
+
+. The `GenerateHeader` rule can now be used in the sources of your project, for example:
++
+[source,txt]
+----
+GenerateHeader xml/config.xml : include/config.hpp ;
+
+exe myProgram
+    : # sources
+    [ glob src/*.cpp ]
+    include/config.hpp
+    # libraries
+    /boost/filesystem//boost_filesystem
+    /boost/system//boost_system
+    ;
+----
++
+The `exe` rule describes how to build an executable. In this case, it is saying to build an executable named `myProgram` from the source files in `src/` and the generated config.hpp. It also lists the Boost libraries that the program depends on.
++
+The `GenerateHeader` rule will ensure that config.xml is transformed into config.hpp before `myProgram` is compiled. This means that changes to config.xml will cause config.hpp to be regenerated, which in turn will trigger a recompilation of any source files that include config.hpp. To execute this custom build step, navigate to your project's root directory and run `b2`.
+
+== Multi-platform Support
+
+Let’s say you are developing an application that needs to run on multiple platforms, including Linux, Windows, and MacOS. Your app uses boost:asio[] for asynchronous networking, boost:thread[] for multithreading, and boost:filesystem[] for file handling:
+
+. Before you can use Boost libraries, they might need to be compiled. You can use Boost.Build to compile the libraries by navigating to the root directory of Boost and running `b2`. If you want to compile only the required libraries, use the `--with-` option:
+
++
+[source,txt]
+----
+./b2 --with-thread --with-filesystem --with-system
```

Comment:
This is really important and should be mentioned first, if not in the getting started.

Reply:
moved to the top